libraries/standard/http/src/http_client.c

```diff
@@ -247,40 +308,203 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
-                                 "did not send the required bytes: Required Bytes=%d"
-                                 ", Sent Bytes=%d.",
-                                 transportStatus );
-            returnStatus = HTTP_NETWORK_ERROR;
+            IotLogErrorWithArgs( "Received partial response from transport ",
+                                 "recv(): ResponseSize=%d, TotalBufferSize=%d",
+                                 totalReceived,
+                                 responseBufferLen - totalReceived );
+            returnStatus = HTTP_PARTIAL_RESPONSE;
         }
     }
+    else
+    {
+        /* Empty else for MISRA 15.7 compliance. */
+    }
 
     return returnStatus;
 }
 
-/*-----------------------------------------------------------*/
-
-static HTTPStatus_t _receiveHttpResponse( const HTTPTransportInterface_t * pTransport,
-                                          HTTPResponse_t * pResponse )
+static HTTPStatus_t _receiveAndParseHttpResponse( const HTTPTransportInterface_t * pTransport,
+                                                  HTTPResponse_t * pResponse )
 {
-    /* TODO: Receive the HTTP response with parsing. */
-    return HTTP_SUCCESS;
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    size_t totalReceived = 0;
+    size_t currentReceived = 0;
+    HTTPParsingContext_t parsingContext = { 0 };
+    uint8_t shouldRecv = 0;
+
+    assert( pTransport != NULL );
+    assert( pTransport->recv != NULL );
+    assert( pResponse != NULL );
+
+    /* Initialize the parsing context. */
+    returnStatus = _HTTPClient_InitializeParsingContext( &parsingContext,
+                                                         pResponse->pHeaderParsingCallback );
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        shouldRecv = 1;
+    }
+
+    while( shouldRecv == 1 )
+    {
+        /* Receive the HTTP response data into the pResponse->pBuffer. */
+        returnStatus = _receiveHttpResponse( pTransport,
+                                             pResponse->pBuffer + totalReceived,
+                                             pResponse->bufferLen - totalReceived,
+                                             &currentReceived );
+
+        if( returnStatus == HTTP_SUCCESS )
+        {
+            if( currentReceived > 0 )
+            {
+                totalReceived += currentReceived;
+                /* Data is received into the buffer and must be parsed. */
+                returnStatus = _HTTPClient_ParseResponse( &parsingContext,
+                                                          pResponse->pBuffer + totalReceived,
+                                                          currentReceived );
+            }
+        }
+
+        /* While there are no errors in the transport recv or parsing, we received
+         * data over the transport, the response message is not finished, and
+         * there is room in the response buffer. */
+        shouldRecv = ( uint8_t ) ( ( returnStatus == HTTP_SUCCESS ) &&
```

Comment:
Why not keep this expression in the above happy-path if-case? You wouldn't need the `( returnStatus == HTTP_SUCCESS )` condition

Reply:
I would still need to check ( returnStatus == HTTP_SUCCESS ) since the loop needs to stop on an error returned from _HTTPClient_ParseResponse().

Reply:
By putting this at the end we setup the next looping conditions clearly.